site/content/blog/2025-02-17-portswigger-lab-walkthrough-broken-brute-force-protection-ip-block/index.md

```diff
@@ -0,0 +1,792 @@
+
+```js
+var payload = usernames[count] + MID + passwords[count];
+```
+
+All functions, except the `log()` function, are part of a required interface that must be implemented for Payload Generator scripts. Most script types have specific interfaces that must be implemented to run successfully. You’ll usually run into an `Incorrect Implementation` error when attempting to use a script that does not implement the expected interface correctly:
+
+&nbsp;
+![implementation-error.png](images/implementation-error.png)
+&nbsp;
+
+> [!NOTE]
+> Refer to the default templates for the interfaces required for each script type. You can find these in the Templates section of the Scripts tab. They include documentation comments describing each function and its parameters. You can also create new scripts from the templates. 
+
+You can save the file now. Also, check that the script is enabled, as disabled scripts won’t be available for selection when needed. If you missed the *Enable* checkbox earlier, you can right-click the filename in the *Scripts* tab and select *Enable Script(s)*.
+
+#### Step 3: Add the Payload
+
+Now, we can bring up the Fuzzer dialog. Right-click the POST request we were working on in the Request Editor and select *Fuzz*. Alternatively, right-click the original captured POST request from the *Sites* tab and select *Attack* > *Fuzz*.
+
+In the *Fuzzer Locations* tab, we’ll select the position we want to fuzz and add our payload. The fuzz position (or location) here, as discussed above, is most of the body starting from the username value. Highlight as shown to activate the *Add* button:
+
+&nbsp;
+![fuzz_username_password.png](images/fuzz_username_password.png)
+&nbsp;
+
+After you open the *Add Payload* dialog, you can add the Payload Generator script from the previous step. The default payload generator type is “Strings”. Change this to “Script” and select the appropriate script:
+
+&nbsp;
+![add-payload-generator-script.png](images/add-payload-generator-script.png)
+&nbsp;
+
+Save the payload and return to the main Fuzzer dialog. Then, run the Fuzzer with *Start Fuzzer*.
+
+> [!IMPORTANT]
+> You might notice that even though the payloads are generated correctly (as the logs in the script output panel show), the requests are sent in the wrong order. This causes an unexpected IP block early on. As a result, most of requests are rejected with the message: `You have made too many incorrect login attempts. Please try again in 1 minute(s).`
+>
+> This is due to the Fuzzer’s concurrency settings. You can enable sequential execution by setting the number of execution threads to 1. However, this slows down the Fuzzer a great deal. An alternative is to throttle requests with a small delay of 100–200 milliseconds. See [Options Fuzzer screen](https://www.zaproxy.org/docs/desktop/addons/fuzzer/options/) for more detail.
+
```

Comment:
@kingthorin Just marking out sections where I'd especially appreciate your help whenever you can take a look. I tried to make sure the ZAP-specific details are accurate, but I'm not too certain about some parts.